### DISCUSSION TITLE
How to return multiple errors without data?

### DISCUSSION THREAD

**Discussion by kamilgregorczyk** [ExpediaGroup/graphql-kotlin#1046] [Q&A] [answered]
Hi everyone, I'm looking for a possibility of returning a list of errors in a proper graphqls way (without unions between success/failure responses).

I started using the example provided by the doc but the data field is required, not providing it or providing a null returns an additional error on graphql

```
DataFetcherResult.newResult<LoginResult>()
                .data(null)
                .error(
                    newErrorException()
                        .message("test message")
                        .build()
                ).build()
                ```
                
```

```
"The field at path '/login' was declared as a non null type, but the code involved in retrieving data has wrongly returned a null value.  The graphql specification requires that the parent field be set to null, or if that is non nullable that it bubble up null to its parent and so on. The non-nullable type is 'LoginResult' within parent type 'Mutation'",
      "path": [
        "login"
      ]
 ```
      
      
Is there any good way of returning just errors during mutations which don't require data structure? as I said, I managed to get it working via unions but I prefer a solution similar to DataFetcherResult

**Comment by dariuszkuc** [comment]
If the top level field is marked as non-nullable then attempting to return a null value is a critical error as it violates the type safety guarantee of the GraphQL. If you want to return `null` for your data you have to make those fields nullable.

**Reply by kamilgregorczyk** [reply]
The thing is I don't want to return any data at all, I'd like to return multiple errors to the caller in a graphqls way, when I throw an exception I get such structure:

```

{
  "errors": [
    {
      "message": "meessage",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "path": [
        "login",
        "email"
      ]
    }
  ]
}
```

but this returns one error as only one exception was thrown, now instead of one item in the `errors` array I'd like to put multiple ones

**Reply by dariuszkuc** [reply]
You can return multiple errors but again your response has to be spec compatible - if you attempt to return `null` for non-nullable field it will fail validation which takes precedence and (most likely) removes all other errors.

**Reply by kamilgregorczyk** [reply]
so there is no way of returning error without data like it happens when an exception is thrown ?

**Comment by smyrick** [comment, marked as answer]
@kamilgregorczyk Try not setting `data` at all

```kotlin
DataFetcherResult.newResult<LoginResult>()
    .error(
        newErrorException()
            .message("test message")
            .build()
    ).build()
```

**Reply by kamilgregorczyk** [reply]
Also getting the same error

**Reply by smyrick** [reply]
This works for me

```kotlin
fun errorsOnly(): DataFetcherResult<String> {
        val error = KotlinGraphQLError(RuntimeException("returning error"), listOf(SourceLocation(1, 1)), ResultPath.rootPath().toList())
        return DataFetcherResult.newResult<String>()
            .error(error)
            .build()
    }
```

![Screen Shot 2021-02-12 at 10 24 07 AM](https://user-images.githubusercontent.com/2446877/107807110-869ed480-6d1c-11eb-822f-5ff85d3839b2.png)

**Reply by kamilgregorczyk** [reply]
Yes, and as you can see it returns 2 errors, one the one you wanted and 2nd one complaining about not being able to parse data 
<img width="2095" alt="Zrzut ekranu 2021-02-12 o 21 10 54" src="https://user-images.githubusercontent.com/9266336/107818049-4b7dbf00-6d77-11eb-897a-5d59ca3b87d0.png">

**Reply by smyrick** [reply]
Correct, because as it states in the error, you didn't return data, so it is an error.

**Reply by kamilgregorczyk** [reply]
Yes, and I'm looking for a solution to return multiple errors without data and without any exceptions saying that my data Is missing :)

**Reply by smyrick** [reply]
The way to solve that would be to either: make the root field selection nullable so it is a valid response if you just return null or to not return a GraphQL response and instead throw an HTTP error.

**Reply by kamilgregorczyk** [reply]
Making fields nullable is not a good solution, also throwing and HTTP error will result in only one message.

 I tried to find somewhere where expeditions are caught to maybe write my own processor for my own exception which could hold multiple messages but I couldn’t find anything that’s open for extension

**Reply by smyrick** [reply]
If you think that this behavior should be modified, it is handled in the underlying [graphql-java](https://www.graphql-java.com/) library.

**Reply by kamilgregorczyk** [reply]
thanks, will go there for help
